dipy/reconst/ivim.py

```diff
@@ -0,0 +1,383 @@
+#!/usr/bin/python
+""" Classes and functions for fitting ivim model """
+from __future__ import division, print_function, absolute_import
+import warnings
+import functools
+import numpy as np
+from dipy.core.optimize import Optimizer
+
+from scipy.optimize import leastsq
+
+from dipy.core.gradients import gradient_table
+from dipy.reconst.base import ReconstModel
+from dipy.reconst.dti import _min_positive_signal, apparent_diffusion_coef
+from dipy.reconst.dti import TensorModel, mean_diffusivity
+from .vec_val_sum import vec_val_vect
+from dipy.core.sphere import Sphere
+
+
+def ivim_function(params, bvals):
+    """The ivim function
+    Parameters
+    ----------
+    params : array
+             The IVIM parameters [S0, f, D_star, D]
+
+    bvals  : array
+             bvalues
+    """
+    S0, f, D_star, D = params
+    S = S0 * (f * np.exp(-bvals * D_star) + (1 - f) * np.exp(-bvals * D))
+    return S
+
+
+def _ivim_error(params, bvals, signal):
+    """Error function to be used in fitting the model
+    """
+    return (signal - ivim_function(params, bvals))
+
+
+class IvimModel(ReconstModel):
+    """Ivim model
+    """
+
+    def __init__(self, gtab, split_b=200.0, min_signal=None):
+        """
+        An IVIM model
+
+        Parameters
+        ----------
+        gtab : GradientTable class instance
+        split_b : split the data at this b-value for a two stage fit
+
+        References
+        ----------
+        .. [1] Le Bihan, Denis, et al. "Separation of diffusion
+                   and perfusion in intravoxel incoherent motion MR
+                   imaging." Radiology 168.2 (1988): 497-505.
+        .. [2] Federau, Christian, et al. "Quantitative measurement
+                   of brain perfusion with intravoxel incoherent motion
+                   MR imaging." Radiology 265.3 (2012): 874-881.
+        """
+        ReconstModel.__init__(self, gtab)
+        self.split_b = split_b
+        self.min_signal = min_signal
+        if self.min_signal is not None and self.min_signal <= 0:
+            e_s = "The `min_signal` key-word argument needs to be strictly"
+            e_s += " positive."
+            raise ValueError(e_s)
+
+    def fit(self, data, mask=None, x0=None, fit_method="one_stage", routine="minimize",
+            jac=None, bounds=((0, None), (0, 1.), (0, 1.), (0, 1.)), tol=1e-25, algorithm='L-BFGS-B',
+            gtol=1e-25, ftol=1e-25, eps=1e-15, maxiter=1000):
```

Comment:
But some of the inputs are specific to the fitting. For example, the fitting method or routine, the algorithm to use, jacobian, the x0 to use. As the data changes won't these change too ?